src/libraries/System.Security.Cryptography.Cose/src/System/Security/Cryptography/Cose/CoseMultiSignMessage.cs

```diff
@@ -0,0 +1,434 @@
+            encodedLength += CoseHelpers.GetIntegerEncodedSize(Signatures.Count);
+
+            foreach (CoseSignature signature in Signatures)
+            {
+                encodedLength += CoseHelpers.SizeOfArrayOfLessThan24;
+                encodedLength += CoseHelpers.GetByteStringEncodedSize(signature._encodedSignProtectedHeaders.Length);
+                encodedLength += CoseHeaderMap.ComputeEncodedSize(signature.UnprotectedHeaders);
+                encodedLength += CoseHelpers.GetByteStringEncodedSize(signature._signature.Length);
+            }
+
+            return encodedLength;
+        }
+
+        public override bool TryEncode(Span<byte> destination, out int bytesWritten)
+        {
+            ThrowIfDuplicateLabels(ProtectedHeaders, UnprotectedHeaders);
+            if (Signatures.Count < 1)
+            {
+                throw new CryptographicException(SR.MultiSignMessageMustCarryAtLeastOneSignature);
+            }
+
+            if (destination.Length < GetEncodedLength())
+            {
+                bytesWritten = 0;
+                return false;
+            }
+
+            var writer = new CborWriter();
+
+            if (_isTagged)
+            {
+                writer.WriteTag(MultiSignTag);
+            }
+
+            writer.WriteStartArray(MultiSignArrayLength);
+
+            writer.WriteByteString(_protectedHeaderAsBstr);
+
+            CoseHelpers.WriteHeaderMap(destination, writer, UnprotectedHeaders, isProtected: false, null);
+
+            CoseHelpers.WriteContent(writer, Content.GetValueOrDefault().Span, !Content.HasValue);
+
+            writer.WriteStartArray(Signatures.Count);
+            foreach (CoseSignature signature in Signatures)
+            {
+                ThrowIfDuplicateLabels(signature.ProtectedHeaders, signature.UnprotectedHeaders);
+                writer.WriteStartArray(CoseSignatureArrayLength);
+                writer.WriteByteString(signature._encodedSignProtectedHeaders);
+
+                CoseHelpers.WriteHeaderMap(destination, writer, signature.UnprotectedHeaders, false, null);
+
+                writer.WriteByteString(signature._signature);
+                writer.WriteEndArray();
+            }
+
+            writer.WriteEndArray();
+            writer.WriteEndArray();
+
+            bytesWritten = writer.Encode(destination);
+            Debug.Assert(bytesWritten == GetEncodedLength());
+
+            return true;
+        }
+
+        public void AddSignature(CoseSigner signer, byte[]? associatedData = null)
+            => AddSignature(signer, associatedData.AsSpan());
+
+        public void AddSignature(CoseSigner signer, ReadOnlySpan<byte> associatedData)
+        {
+            if (signer == null)
+            {
+                throw new ArgumentNullException(nameof(signer));
+            }
+
+            if (IsDetached)
+            {
+                throw new CryptographicException(SR.Sign1VerifyContentWasDetached);
```

Comment:
Does this mean we need to reconsider AddSignature as AddSignatureEmbedded and AddSignatureDetached?

Reply:
Yes, and AddSignatureDetachedAsync.